### ISSUE TITLE
Understanding the ordering of migrations

### ISSUE THREAD

**Issue by nicwest** [weavejester/ragtime#76]
I'm probably being dense, but I'm struggling to get migrations to execute in the correct order. here is my simple project:

```clj
;src/testragtime/core.clj
(ns testragtime.core
  (:gen-class)
  (:require [ragtime.jdbc :as jdbc]
            [ragtime.repl :as repl]))

(def migration-config
  {:database (jdbc/sql-database {:connection-uri "jdbc:postgresql://localhost:5432/testdb?user=nic&password=password"})
   :migrations (jdbc/load-resources "migrations")})

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))
```
```clj
;project.clj
(defproject testragtime "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.6.0"]
                 [ragtime "0.4.1"]
                 [postgresql "9.3-1102.jdbc41"]]
  :main ^:skip-aot testragtime.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})
```

```sql
-- resources/migrations/001-thing.up.sql
CREATE TABLE thing (id int);
-- resources/migrations/001-thing.down.sql
DROP TABLE thing;

-- resources/migrations/002-stuff.up.sql
ALTER TABLE thing ADD stuff date;
-- resources/migrations/002-stuff.down.sql
ALTER TABLE thing DROP stuff;

-- resources/migrations/003-misc.up.sql
CREATE TABLE misc (id int);
-- resources/migrations/003-misc.down.sql
DROP TABLE misc;
```

Then in the repl:

```
testragtime.core=> migration-config
{:database #ragtime.jdbc.SqlDatabase{:db-spec {:connection-uri "jdbc:postgresql://localhost:5432/testdb?user=nic&password=nic"}, :migrations-table "ragtime_migrations"}, :migrations ({:id "003-misc", :up #<jdbc$sql_migration$fn__739 ragtime.jdbc$sql_migration$fn__739@61f4bdad>, :down #<jdbc$sql_migration$fn__741 ragtime.jdbc$sql_migration$fn__741@15e04bdb>} {:id "002-stuff", :up #<jdbc$sql_migration$fn__739 ragtime.jdbc$sql_migration$fn__739@38942215>, :down #<jdbc$sql_migration$fn__741 ragtime.jdbc$sql_migration$fn__741@549adb8>} {:id "001-thing", :up #<jdbc$sql_migration$fn__739 ragtime.jdbc$sql_migration$fn__739@282c0dbe>, :down #<jdbc$sql_migration$fn__741 ragtime.jdbc$sql_migration$fn__741@643a498c>})}
testragtime.core=> (repl/migrate migration-config)
Applying 003-misc
Applying 002-stuff

PSQLException ERROR: relation "thing" does not exist  org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse (QueryExecutorImpl.java:2198)
```

I would expect it to execute 001-things then 002-stuff then 003-misc, but it seems to be doing it in reverse order.

strangely if I rename 001-things.\*.sql to 004-things.\*.sql, it has no effect at all.

I'm clearly missing something here.

**Comment by weavejester** [comment]
No, you're not missing anything. A `sort` was missed out of the code, but nothing broke because the system it was tested on happened to have the right order by default. I've added a fix.

**Comment by nicwest** [comment]
wow super quick thanks alot :D

**Comment by weavejester** [comment]
I'm going to release a 0.4.2-SNAPSHOT in a minute. Do you mind testing it out to see if it fixes your issue?

**Comment by nicwest** [comment]
sure!

**Comment by weavejester** [comment]
Okay, done. Let me know if 0.4.2-SNAPSHOT solves the issue. If it does, I'll release 0.4.2.

**Comment by nicwest** [comment]
yup LGTM!

**Comment by weavejester** [comment]
Okay, 0.4.2 released. Sorry for the issue.

**Comment by nicwest** [comment]
thanks for the speedy response, keep up the good work!

**Comment by csummers** [comment]
I don't quite think this is resolved.  I'm still seeing unordered migration issues in 0.5.0.  

Adding a ```(sort-by :id)``` to the end of that threading pipeline in ```load-all-files``` seems to fix it.

**Comment by weavejester** [comment]
Are you certain that a `lein clean` doesn't fix your issue, @csummers? `load-all-files` has a `sort-by` in it already, and it passes all the tests, so in theory there shouldn't be anything wrong with it...

**Comment by csummers** [comment]
I'm fairly certain as I'm locally installing a patched jar to test.  I think the issue is the ```group-by``` in the ```for``` of ```load-files```.  group-by returns a hashmap, so there are no order guarantees, and this occurs after the sort occurs.

**Comment by weavejester** [comment]
The `group-by` is by file extension, so all the files with the same extension should be in order. However, if you're mixing `.edn` and `.sql` files there would potentially be problems.

**Comment by csummers** [comment]
I'm referring to the ```group-by``` in ```load-files``` called by ```(mapcat load-files)``` https://github.com/weavejester/ragtime/blob/d5a48083840c364b2a65c15bbe36f3634ab50eda/ragtime.jdbc/src/ragtime/jdbc.clj#L123

However, I'll try to come up with a reproducible test case to demonstrate the issue.

**Comment by weavejester** [comment]
Ah, I think you're right. The order would get mixed up in the first `group-by` of `load-files`, and the tests don't check enough files to catch that.

**Comment by csummers** [comment]
Thanks for the help!  I'm really liking ragtime.

**Comment by weavejester** [comment]
This *should* be fixed in 0.5.1 now. My apologies for the issue. Could you let me know if 0.5.1 solves your problem this time?

**Comment by csummers** [comment]
Yes, 0.5.1 fixes this bug for me.  Thanks again.
